extensions/objects/templates/image-editor.component.ts

```diff
@@ -0,0 +1,1155 @@
+        this.data.metadata.savedImageFilename)
+    ) {
+      this.imageLocalStorageService.deleteImage(
+        this.data.metadata.savedImageFilename);
+    }
+    this.data = {
+      mode: this.MODE_EMPTY,
+      metadata: {},
+      crop: true
+    };
+    this.imageResizeRatio = 1;
+    this.invalidTagsAndAttributes = {
+      tags: [],
+      attrs: []
+    };
+    this.validityChange.emit({empty: false});
+  }
+
+  validate(data: FilepathData): boolean {
+    const isValid = data.mode === this.MODE_SAVED &&
+      data.metadata.savedImageFilename &&
+      data.metadata.savedImageFilename.length > 0;
+    return isValid;
+  }
+
+  isUserCropping(): boolean {
+    const dimensions = this.calculateTargetImageDimensions();
+    const cropWidth = this.cropArea.x2 - this.cropArea.x1;
+    const cropHeight = this.cropArea.y2 - this.cropArea.y1;
+    return cropWidth < dimensions.width || cropHeight < dimensions.height;
+  }
+
+  onMouseMoveOnImageArea(e: MouseEvent): void {
+    e.preventDefault();
+
+    const coords = this.getEventCoorindatesRelativeToImageContainer(e);
+
+    if (this.userIsDraggingCropArea) {
+      this.handleMouseMoveWhileDraggingCropArea(coords.x, coords.y);
+    } else if (this.userIsResizingCropArea) {
+      this.handleMouseMoveWhileResizingCropArea(coords.x, coords.y);
+    } else {
+      this.updatePositionWithinCropArea(coords.x, coords.y);
+    }
+
+    this.mouseLastKnownCoordinates = { x: coords.x, y: coords.y };
+  }
+
+  onMouseDownOnCropArea(e: MouseEvent): void {
+    e.preventDefault();
+    const coords = this.getEventCoorindatesRelativeToImageContainer(e);
+    const position = this.mousePositionWithinCropArea;
+
+    if (position === this.MOUSE_INSIDE) {
+      this.lastMouseDownEventCoordinates = { x: coords.x, y: coords.y };
+      this.cropAreaXWhenLastDown = this.cropArea.x1;
+      this.cropAreaYWhenLastDown = this.cropArea.y1;
+      this.userIsDraggingCropArea = true;
+    } else if (position !== null) {
+      this.lastMouseDownEventCoordinates = { x: coords.x, y: coords.y };
+      this.userIsResizingCropArea = true;
+      this.cropAreaResizeDirection = position;
+    }
+  }
+
+  onMouseUpOnCropArea(e: MouseEvent): void {
+    e.preventDefault();
+    this.userIsDraggingCropArea = false;
+    this.userIsResizingCropArea = false;
+  }
+
+  getMainContainerDynamicStyles(): string {
+    const width = this.OUTPUT_IMAGE_MAX_WIDTH_PX;
+    return 'width: ' + width + 'px';
+  }
+
+  getImageContainerDynamicStyles(): string {
+    if (this.data.mode === this.MODE_EMPTY) {
+      return 'border: 1px dotted #888';
+    } else {
+      return 'border: none';
+    }
+  }
+
+  getToolbarDynamicStyles(): string {
+    if (this.isUserCropping()) {
+      return 'visibility: hidden';
+    } else {
+      return 'visibility: visible';
+    }
+  }
+
+  getCropButtonBarDynamicStyles(): string {
+    return 'left: ' + this.cropArea.x2 + 'px;' +
+      'top: ' + this.cropArea.y1 + 'px;';
```

Comment:
Can we use some kind of string formatting for this?
```suggestion
    return `left: {this.cropArea.x2}px; top: {this.cropArea.y1}px;`;
```

Reply:
Ditto in other places